pytensor/graph/utils.py

```diff
@@ -420,3 +431,111 @@ def toposort(prereqs_d):
             "some orderings contain invalid elements."
         )
     return seq
+
+
+def _graph_replace(
+    outputs: Sequence["Variable"], replace: Dict["Variable", "Variable"]
+) -> Tuple[List["Variable"], Dict["Variable", "Variable"]]:
+    """Replace variables in ``outputs`` by ``replace`` in a single pass.
+
+    Parameters
+    ----------
+    outputs: Sequence[Variable]
+        Output graph
+    replace: Dict[Variable, Variable]
+        Replace mapping
+
+    Returns
+    -------
+    List["Variable"]
+        Output graph with subgraphs replaced
+    Dict["Variable", "Variable"]
+        Replacements that were not applied
+    """
+    from pytensor.graph.basic import Constant, condition_subset
+    from pytensor.graph.fg import FunctionGraph
+
+    # collect minimum graph inputs which is required to compute outputs
+    # and depend on replacements
+    # additionally remove constants, they do not matter in clone get equiv
+    conditions = [
+        c for c in condition_subset(outputs, replace) if not isinstance(c, Constant)
+    ]
+    # for the function graph we need the clean graph where
+    # inputs do not have owners
+    # this is exactly the reason to clone conditions
+    equiv = {c: c.clone(name=f"i-{i}") for i, c in enumerate(conditions)}
```

Comment:
Curious why you ended up with this approach instead of the `clone_get_equiv` with `memo={c: c}`?

Reply:
Clone get equivalent is called the same way in fg

Reply:
https://github.com/pymc-devs/pytensor/blob/d78e1a1ad7b4b2f4d13d0c3043ea79ca53aa24d0/pytensor/graph/fg.py#L112

Reply:
Neat, I didn't know it ;)

Reply:
Still why doesn't `memo={c: c}` work? Why do you need to erase the graph upstream?

Reply:
Because Function graph expects `c` does not have owner